olp-cpp-sdk-core/src/cache/DefaultCacheImpl.cpp

```diff
@@ -529,17 +532,84 @@ uint64_t DefaultCacheImpl::MaybeEvictData(leveldb::WriteBatch& batch) {
   }
 
   const auto start = std::chrono::steady_clock::now();
-  int64_t evicted = 0u;
+  int64_t left_to_evict =
+      mutable_cache_data_size_ -
+      std::llroundl(settings_.max_disk_storage * kMinDiskUsedThreshold);
+  uint64_t evicted = 0u;
   auto count = 0u;
-  const auto min_size = kMinDiskUsedThreshold * settings_.max_disk_storage;
 
+  const auto call_evict_method =
+      [&](std::function<EvictionResult(DefaultCacheImpl*, leveldb::WriteBatch&,
+                                       uint64_t)>
+              evict_method) {
+        EvictionResult eviction_result;
+        std::unique_ptr<leveldb::WriteBatch> eviction_batch = nullptr;
+        uint64_t current_eviction_target;
+
+        do {
+          // At this point left_to_evict can't be below zero.
+          current_eviction_target =
+              static_cast<uint64_t>(left_to_evict) < eviction_portion_
+                  ? left_to_evict
+                  : eviction_portion_;
+          eviction_batch = std::make_unique<leveldb::WriteBatch>();
+          eviction_result =
+              evict_method(this, *eviction_batch, current_eviction_target);
+
+          if (eviction_result.size >= eviction_portion_) {
+            // Apply intermediate batch
+            const auto apply_result =
+                mutable_cache_->ApplyBatch(std::move(eviction_batch));
```

Comment:
I am not sure about this. leveldb can be tricky :-)
AFAIK, leveldb isn't deleting data instantly after applying batch, it just marks keys as deleted. Thus, data is not available anymore, but still stored on the disk. Compaction should free not needed data.
Previously, eviction could happen only when adding new data. leveldb has internal mechanism to trigger the compaction, so cache size is keeping around `max_disk_storage` value.
Lowering cache size probably means user need some free space, so we definitely must trigger the compaction process.
The operation isn't cheap, so not sure what is better - compact whole cache each time `eviction_portion_` removed or only once in the end.
If we will do only 1 compaction after all data is removed, there is no need to split deletion into several batches, because real work done only in the end.

Reply:
Discussed in private. Compaction is added after every ApplyBatch.
Will be tested on a big amount of data. The algorithm may be optimized based on the results.